44.md

```diff
@@ -0,0 +1,166 @@
+NIP-44
+======
+
+Encrypted Payloads (Versioned)
+------------------------------
+
+`optional` `author:paulmillr` `author:staab`
+
+The NIP introduces a new data format for keypair-based encryption. This NIP is versioned to allow multiple algorithm choices to exist simultaneously.
+
+An encrypted payload MUST be encoded as a JSON object. Different versions may have different parameters. Every format has a `v` field specifying its version.
+
+Currently defined encryption algorithms:
+
+- `0x00` - Reserved
+- `0x01` - XChaCha with same key `sha256(ecdh)` per conversation
+
+# Version 1
+
+Params:
+
+1. `nonce`: base64-encoded xchacha nonce
+2. `ciphertext`: base64-encoded xchacha ciphertext, created from (key, nonce) against `plaintext`.
+
+Example:
+
+- Alice's private key: `5c0c523f52a5b6fad39ed2403092df8cebc36318b39383bca6c00808626fab3a`
+- Bob's private key: `4b22aa260e4acb7021e32f38a6cdf4b673c6a277755bfce287e370c924dc936d`
+
+Encrypting the message `hello` from Alice to Bob results in the base-64 encoded tlv payload:
+
+```
+AAEBARgeI8gcP/4mnw3mKgtMvD8aGYUnGBlhopoCBd94Ev9i
+```
+
+# Other Notes
+
+By default in the [libsecp256k1](https://github.com/bitcoin-core/secp256k1) ECDH implementation, the secret is the SHA256 hash of the shared point (both X and Y coordinates). We are using this exact implementation. In NIP-94, unhashed shared point was used.
+
+This encryption scheme replaces the one described in NIP-04, which is not secure. It used bad cryptographic building blocks and must not be used.
+
+# Code Samples
+
+## Javascript
+
+```javascript
+import {base64} from '@scure/base'
+import {randomBytes} from '@noble/hashes/utils'
+import {secp256k1} from '@noble/curves/secp256k1'
+import {sha256} from '@noble/hashes/sha256'
+import {xchacha20} from '@noble/ciphers/chacha'
+
+import {utf8Decoder, utf8Encoder} from './utils.ts'
+
+export const getSharedSecret = (privkey: string, pubkey: string): Uint8Array =>
+  sha256(secp256k1.getSharedSecret(privkey, '02' + pubkey).subarray(1, 33))
+
+export function encrypt(privkey: string, pubkey: string, text: string, v = 1) {
```

Comment:
Should it be `privkey: string, pubkey: string`, or rather `key: Uint8Array`?

Caching an encryption key somewhere is definitely a goal. Using `privkey, pubkey` signature makes it impossible.

Reply:
No, user can't re-derive encryption key for 50 messages. That would consume all the performance. ECDH is slow, like, 500 per sec on M2. They need to derive key once and use the same key for encryption / decryption of all conversation messages, and there can be 1000s.

Reply:
```js
const msgs = await fetchMessages(100);

// original
let key = getSharedSecret(privA, pubB);
msgs.map(msg => decrypt(key, msg)); // pure decryption routine: 100K/sec

// vs your proposal
msgs.map(msg => decrypt(privA, pubB, msg)); // decryption + getSharedSecret: 200x slower
```

This is just one convo. There can be like 20.

Reply:
This sounds like an optimization, couldn't you just lru_cache the `getSharedSecret` function?

Reply:
One frame is best-case 16.6ms, worst-case 8.3ms. Decrypting messages in main thread (when you can't use separate one) would just make UI stuck. Even with a separate thread, decrypting 500 conversation messages could take 2-5 seconds on a mobile phone with your proposal. Forcing users to wait is terrible UX.

Reply:
Got it, yes the base version should just take a key, I'll update that. For stuff like browser extensions, there's no need to expose the key-only version (if the application has a key they can import nostr-tools or whatever), so that interface will look a little different.